Add hotkey for removing selected actors from control groups

dragunoff opened the pull request:
In this PR:
* Add `RemoveFromControlGroup` hotkey (defaults to `Unknown`)
* Add `RemoveFromControlGroup` method to `OpenRA.Game.Selection`

~~The default hotkey is up for discussion but I think it's a good choice because it's on the same row as the numbers, is easy to reach and is functions similarly to how control groups are assigned~~.

Closes #15961

dragunoff committed:
Add hotkey for removing actors from control groups

* Add `RemoveFromControlGroup` hotkey
* Add `RemoveFromControlGroup` method to `OpenRA.Game.Selection`

pchote commented:
I'd like to see other people's opinions on this, but IMO we shouldn't add this - I don't think any of the original games had a specific hotkey for this, and the same function can be achieved by adding the unit to a different group.

Every new hotkey type adds an overhead for players to remember, and a maintenance burden / opportunity cost for us (reducing the number of keys available for other things).  Also note that <kbd>`</kbd> in particular moves around on various international keyboard layouts, and when paired with <kbd>Cmd</kbd> on macOS (all <kbd>Ctrl</kbd> shortcuts are remapped to <kbd>Cmd</kbd>) it doesn't seem to work.

MustaphaTR commented:
I dunno if our shortcut code actually look for the keys, or the place on the keyboard. But just to note, in Hungarian Keyboard, there is <kbd>0</kbd> to the left of <kbd>1</kbd>. Maybe CTRL + Backspace can be a better idea if this ever gets in.

matjaeck commented:
Most people will have to use both hands to hit these keys. If we look back at the original issue, time and control have been the main factors that motivated the author to ask for this feature. In his opinion, re-assigning the control-group would create a greater "mess" than removing certain units from a control-group. I know that these settings could be changed to a more ergonomic setup by the player, however these defaults alone make me doubt if there is any additional value in this change.

The author raised another point that is more interesting IMO: Removing units from control-groups when they are sent to repair. It is the underlying problem this feature request tries to address. You have to remove those units from a control group, that you want to send to repair. Otherwise any order given to the control-group will "overwrite" the repair order.

IMO, there is no problem with assigning control-groups (which includes the removal of units from control-groups). However - and this would need to be discussed separately - I would support a change that removes units (temporarily?) from control groups, when they are send to repair.

MustaphaTR is right that a unit can only be in one control group. This mean that you already have a very fast and more comfortable solution to remove single units from conrol-groups by assigning a "dummy" control-group to them. In the light of this, I don't see additional value in this change that justifies the costs pchote already mentioned.

rautamiekka commented:
Been dreaming about this for a very long time. Not a single game allows to drop an unit from a group, and it's indeed more of a mess to reassign than drop since those units will still involve themselves when they shouldn't.

dragunoff commented:
@pchote Those are all valid points and things that I've been thinking about as well. I agree we need more opinions here. Here's my opinion:

**Overhead for players**
I think this is not a big problem as players don't bother remembering all of the hotkeys but only the ones that they find useful. For example there a lot of keys that I don't use - viewport bookmarks, production facilities cycle and player stance toggle (I wish there was a way to unbind them but that's another issue). 

Any piece of software with a complex UI will probably have a lot of hotkeys (think of photo editing suites, softrawe IDEs, etc.) and novice users will perform most tasks first with the mouse, gradually learning the hotkeys and deciding which ones make sense to their workflow (playstyle).

**Maintenance cost**
This is a fairly simple hotkey.

**The need for this hotkey**
This is the most important part of the question... The most common use-case is for units that are ordered to repair. Other use-cases will depend on player style - a few from my experience: splitting a small part of a grouped army, managing air units (I tend to use 2-3 control groups for aircraft). 

Perhaps this is not sufficient to justify the existence of such a hotkey. And as mentioned there is a work-around (adding to a dummy group) but to me that feels awkward.

abcdefg30 reviewed (commented):
I agree with Mustapha that `ctrl` + `backspace` might be a better default. (As it should work across all systems? Plus backspace fits to removing an item.)
I think this feature can be really useful. (Especially with aircraft splitting up for rearming.)

dragunoff commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/RemoveFromControlGroupsHotkeyLogic.cs:
```diff
@@ -0,0 +1,47 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
```
Fixed.

dragunoff commented:
Rebased and updated.

* Default hotkey changed to `Ctrl + BACKSPACE`
* Shorten hotkey description (it almost fits in the UI now, I'm open to suggestions for better wording)

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/RemoveFromControlGroupsHotkeyLogic.cs:
```diff
@@ -0,0 +1,47 @@
+				.ToArray();
+
+			foreach (var a in selectedActors)
+			{
```
Please drop the braces around the single-line statement.

obrakmann commented on mods/common/hotkeys/game.yaml:
```diff
@@ -18,6 +18,10 @@ SelectUnitsByType: W
 	Types: World
 
+RemoveFromControlGroups: BACKSPACE Ctrl
+	Description: Remove from control groups
```
Please make it "Remove from control group", otherwise the text overflows in the dialog. Also please leave the default hotkey empty, as a compromise. :+1: after that.

pchote replied:
Yeah, i'm happy to let my objections slide if this becomes an option feature that players enable themselves.

dragunoff replied:
Changes applied.

dragunoff commented:
Rebased and applied requested changes.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/25bad66d3aafc45bfa0e35186cd1e7c93da7d66a)
